h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,189 @@
+package ai.h2o.targetencoding;
+
+import water.MRTask;
+import water.fvec.CategoricalWrappedVec;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+
+class BroadcastJoinForTargetEncoder {
+
+  static class FrameWithEncodingDataToArray extends MRTask<FrameWithEncodingDataToArray> {
+    
+    /**
+     * As it is less likely that we have more folds than categorical levels It is more preferable to keep `num` and `den` in a separate (but adjacent) rows of the 2D array and not in the same row with shifting.
+     *     For _foldColumnIdx == -1
+     *        _encodingDataPerNode[0] will be storing numerators
+     *        _encodingDataPerNode[1] will be storing denominators
+     *     For _foldColumnIdx != -1
+     *        _encodingDataPerNode[2k-2] will be storing numerators, 
+     *        _encodingDataPerNode[2k-1] will be storing denominators, where k is a current fold ; folds = {1,2,3...k}
+     */
+    int[][] _encodingDataPerNode = null;
+    
+    int _categoricalColumnIdx, _foldColumnIdx, _numeratorIdx, _denominatorIdx, _cardinalityOfCatCol, _maxFoldValue;
+
+    FrameWithEncodingDataToArray(int categoricalColumnIdx, int foldColumnId, int numeratorIdx, int denominatorIdx, int cardinalityOfCatCol, int maxFoldValue) {
+      _categoricalColumnIdx = categoricalColumnIdx;
+      _foldColumnIdx = foldColumnId;
+      _numeratorIdx = numeratorIdx;
+      _denominatorIdx = denominatorIdx;
+      _cardinalityOfCatCol = cardinalityOfCatCol;
+      _maxFoldValue = maxFoldValue;
+
+      if(foldColumnId == -1) {
+        _encodingDataPerNode = new int[2][cardinalityOfCatCol];
+      } else {
+        assert maxFoldValue >= 1 : "There should be at leas two folds in the fold column";
+        assert cardinalityOfCatCol > 0 : "Cardinality of categ. column should be greater than zero";
+        _encodingDataPerNode = new int[(maxFoldValue + 1) * 2][cardinalityOfCatCol];
+      }
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      Chunk categoricalChunk = cs[_categoricalColumnIdx];
+      Chunk numeratorChunk = cs[_numeratorIdx];
+      Chunk denominatorChunk = cs[_denominatorIdx];
+      for (int i = 0; i < categoricalChunk.len(); i++) {
+        int levelValue = (int) categoricalChunk.at8(i);
+        int[] arrForNumerators = null;
+        int[] arrForDenominators = null;
+
+        synchronized (_encodingDataPerNode) { // Sync on whole array as there is no way to sync on two objects ( two rows of 2D array)
```

Comment:
Going to consider to rewrite this. If we were storing numerators and denominators in one row we would be able to sync on single row of the 2D array. But for that we will have to "share" row and maximum number of categorical levels will be around (2^31-1) / 2 instead of (2^31-1). Which is probably quite fine.